src/MISC/fix_eco_force.cpp

```diff
@@ -0,0 +1,642 @@
+  u_0 = atof(arg[3]);
+  sign = (u_0 >= 0.0 ? 1 : -1);
+  eta = atof(arg[4]);
+  r_cut = atof(arg[5]);
+
+  // read reference orientations from file
+  // work on rank 0 only
+  int n = strlen(arg[6]) + 1;
+  dir_filename = new char[n];
+  strcpy(dir_filename, arg[6]);
+  if (me == 0) {
+    char line[IMGMAX];
+    char *result;
+    int count;
+
+    FILE *infile = fopen(dir_filename, "r");
+    if (infile == NULL) error->one(FLERR, "Fix eco/force file open failed");
+    for (int i = 0; i < 6; ++i) {
+      result = fgets(line, IMGMAX, infile);
+      if (!result) error->one(FLERR, "Fix eco/force file read failed");
+      count = sscanf(line, "%lg %lg %lg", &dir_vec[i][0], &dir_vec[i][1], &dir_vec[i][2]);
+      if (count != 3) error->one(FLERR, "Fix eco/force file read failed");
+    }
+    fclose(infile);
+
+    // calculate reciprocal lattice vectors
+    get_reciprocal();
+
+    squared_cutoff = r_cut * r_cut;
+    inv_squared_cutoff = 1.0 / squared_cutoff;
+    half_u = 0.5 * u_0;
+    inv_eta = 1.0 / eta;
+  }
+
+  // initializations
+  MPI_Bcast(&dir_vec[0][0], 18, MPI_DOUBLE, 0, world);                  // communicate direct lattice vectors
+  MPI_Bcast(&reciprocal_vectors[0][0][0], 18, MPI_DOUBLE, 0, world);    // communicate reciprocal vectors
+  MPI_Bcast(&squared_cutoff, 1, MPI_DOUBLE, 0, world);                  // communicate squared cutoff radius
+  MPI_Bcast(&inv_squared_cutoff, 1, MPI_DOUBLE, 0, world);              // communicate inverse squared cutoff radius
+  MPI_Bcast(&half_u, 1, MPI_DOUBLE, 0, world);                          // communicate half potential energy
+  MPI_Bcast(&inv_eta, 1, MPI_DOUBLE, 0, world);                         // communicate inverse threshold
+
+  // set comm size needed by this Fix
+  if (u_0 != 0.0) comm_forward = 110;
+  else comm_forward = 98;
+
+  added_energy = 0.0;                                   // initial energy
+
+  nmax = atom->nmax;
+  nbr = (Nbr *) memory->smalloc(nmax * sizeof(Nbr), "eco/force:nbr");
+  memory->create(order, nmax, 2, "eco/force:order");
+  array_atom = order;
+
+  // zero the array since a variable may access it before first run
+  for (int i = 0; i < atom->nlocal; ++i) order[i][0] = order[i][1] = 0.0;
+}
+
+/* ---------------------------------------------------------------------- */
+
+FixECOForce::~FixECOForce() {
+  memory->destroy(order);
+  memory->sfree(nbr);
+  delete[] dir_filename;
+}
+
+/* ---------------------------------------------------------------------- */
+
+int FixECOForce::setmask() {
+  int mask = 0;
+  mask |= POST_FORCE;
+  mask |= THERMO_ENERGY;
+  mask |= POST_FORCE_RESPA;
+  return mask;
+}
+
+/* ---------------------------------------------------------------------- */
+
+void FixECOForce::init() {
+  // get this processors rank
+  MPI_Comm_rank(world, &me);
+
+  // compute normalization factor
+  if (me == 0) {
+    int neigh;
+    get_norm(neigh);
+    printf("fix eco/force: cutoff=%f norm_fac=%f neighbors=%i\n", r_cut, norm_fac, neigh);
```

Comment:
please never use printf() like this. instead output to either the `screen` or both `screen` and `logfile` FILE pointers, if non-NULL, as it is done throughout the code. Those can be configured to be turned off (and thus any output unlike printf()s).
since the last patch release yesterday, there is also the option to use `utils::logmesg()` to output such messages
and `fmt::format()` or `fmt::printf()` to have a simpler code (please see https://fmt.dev for how to use this feature, or other uses in the current master branch).